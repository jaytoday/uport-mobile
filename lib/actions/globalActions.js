// Copyright (C) 2018 ConsenSys AG
//
// This file is part of uPort Mobile App.
//
// uPort Mobile App is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// uPort Mobile App is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with uPort Mobile App.  If not, see <http://www.gnu.org/licenses/>.
//
import { RESET_DEVICE, RELOAD_DB, PERSIST_DB, FORCE_PERSIST_DB, LOADED_DB } from '../constants/GlobalActionTypes'

export function resetDevice () {
  return { type: RESET_DEVICE }
}

export function reload (db) {
  return {
    type: RELOAD_DB,
    db
  }
}

// trigger persistence of a certain reducer (the key in the global reducer. eg 'uport')
export function persist (reducer) {
  return {
    type: PERSIST_DB,
    reducer
  }
}

// trigger persistence of a certain reducer (the key in the global reducer. eg 'uport')
export function forcePersistDB () {
  return {
    type: FORCE_PERSIST_DB
  }
}

export function loadedDB () {
  return { type: LOADED_DB }
}
