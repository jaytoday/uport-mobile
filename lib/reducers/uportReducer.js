// Copyright (C) 2018 ConsenSys AG
//
// This file is part of uPort Mobile App.
//
// uPort Mobile App is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// uPort Mobile App is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with uPort Mobile App.  If not, see <http://www.gnu.org/licenses/>.
//

import { STORE_IDENTITY, STORE_SUB_ACCOUNT, UPDATE_IDENTITY, SWITCH_IDENTITY, ADD_OWN_CLAIMS, REMOVE_OWN_CLAIM, INCREASE_NONCE, STORE_FUEL_TOKEN,
          SAVE_NONCE, SAVE_META_NONCE, SAVE_BALANCE, SAVE_FUEL_BALANCE, REMOVE_IDENTITY, STORE_EXTERNAL_UPORT, STORE_CONNECTION,
          REMOVE_CONNECTION, STORE_ACTIVITY, UPDATE_ACTIVITY, REMOVE_ACTIVITY, UPDATE_INTERACTION_STATS, OPEN_ACTIVITY,
          STORE_ATTESTATION, REMOVE_ATTESTATION, SAVE_IPFS_PROFILE, ADD_RECOVERY_ADDRESS, SAVE_SEGMENT_ID, STORE_CURRENT_IDENTITY,
          STORE_ENCRYPTION_KEY, STORE_PUTUTU_TOKEN, STORE_KEYCHAIN_LEVEL, STORE_SECURITY_LEVEL } from 'uPortMobile/lib/constants/UportActionTypes'
import { IMPORT_SNAPSHOT } from 'uPortMobile/lib/constants/HubActionTypes'          
import { RESET_DEVICE } from 'uPortMobile/lib/constants/GlobalActionTypes'
import { hashMap, assoc, assocIn, conj, dissoc, disj, empty, get, getIn, updateIn, merge, toClj, inc, set, filter, keys, first, identity, map } from 'mori'
import { sha3_256 } from 'js-sha3'
import { massage } from 'uPortMobile/lib/sagas/stateSaver'
// Add a key/value pair to the current identity
function assocIdentity (state, address, key, value) {
  return assocIn(state, ['byAddress', address, key], value)
}

function findActivity (state, activityId) {
  return first(
          filter(identity, // identity is simply function returning it's parameter, so filter identity removes any value that isn't null or false
            map(address => {
              // console.log(`findActivity ${activityId}, ${address}`)
              const activity = getIn(state, ['byAddress', address, 'activities', activityId.toString()])
              // console.log(toJs(activity))
              if (activity) return {address, activity}
            }, keys(get(state, 'byAddress')))))
}

function uportReducer (state = hashMap(), action) {
  let activityPair
  switch (action.type) {
    case RESET_DEVICE:
      return hashMap()
    case STORE_IDENTITY:
      const existing = getIn(state, ['byAddress', action.identity.address])
      return assoc(
                assocIn(state, ['byAddress', action.identity.address], merge(existing, toClj(action.identity))),
                  'currentIdentity', action.identity.address)
    case STORE_SUB_ACCOUNT:
      return updateIn(state, ['byAddress', action.identity.address], (id) => {
        const newId = toClj(action.identity)
        if (id) {
          return merge(id, dissoc(newId, 'createdAt'))
        } else {
          return newId
        }
      })
    case UPDATE_IDENTITY:
      return updateIn(state, ['byAddress', action.identity], id => {
        if (id) {
          return merge(id, toClj(action.data))
        }
      })
    case SAVE_SEGMENT_ID:
      return assoc(state, 'segmentId', action.segmentId)
    case SWITCH_IDENTITY:
      return assoc(state, 'currentIdentity', action.address)
    case ADD_OWN_CLAIMS:
      // `updateIn` allows you to update a deeply nested piece of data using a function.
      // In this case we find the path into the current identities own claims map and merge the new claims
      // http://swannodette.github.io/mori/#updateIn
      return updateIn(state, ['byAddress', action.address, 'own'],
                            (claims) => merge(claims, toClj(action.claims)))
    case REMOVE_OWN_CLAIM:
      const claims = getIn(state, ['byAddress', action.address, 'own'])
      if (get(claims, action.claimType)) {
        return assocIdentity(state, action.address, 'own', dissoc(claims, action.claimType))
      } else {
        return state
      }
    case STORE_CONNECTION:
      // `updateIn` allows you to update a deeply nested piece of data using a function.
      // In this case we find the path into the current identities own claims map and connection add the address (conj)
      // to the set. If a set doesn't exist we create a new one using set()
      // http://swannodette.github.io/mori/#updateIn
      return updateIn(state, ['byAddress', action.address, 'own', 'connections', action.connectionType],
                            (connections) => conj(set(connections) || set(), action.connection))
    case REMOVE_CONNECTION:
      // `updateIn` allows you to update a deeply nested piece of data using a function.
      // In this case we find the path into the current identities own claims map and connection remove the address (disj)
      // from the connection set.
      // http://swannodette.github.io/mori/#updateIn
      return updateIn(state, ['byAddress', action.address, 'own', 'connections', action.connectionType],
                            (connections) => disj(set(connections), action.connection))
    case ADD_RECOVERY_ADDRESS:
      return updateIn(state, ['byAddress', action.address], (id) => merge(id, toClj({recoveryAddress: action.recoveryAddress, recoveryType: action.recoveryType})))
    case STORE_ENCRYPTION_KEY:
      return updateIn(state, ['byAddress', action.address], (id) => merge(id, toClj({publicEncKey: action.publicEncKey})))
    case STORE_PUTUTU_TOKEN:
      return updateIn(state, ['byAddress', action.address], (id) => merge(id, toClj({pututuToken: action.pututuToken})))
    case STORE_FUEL_TOKEN:
      return assocIdentity(state, action.address, 'fuelToken', action.token)
    case SAVE_NONCE:
      return assocIdentity(state, action.address, 'nonce', action.nonce)
    case SAVE_META_NONCE:
      return assocIdentity(state, action.address, 'metaNonce', { nonce: action.nonce, timestamp: action.timestamp })
    case INCREASE_NONCE:
      return updateIn(state, ['byAddress', action.address, 'nonce'], inc)
    case SAVE_BALANCE:
      return assocIdentity(state, action.address, 'balance', {ethBalance: action.balance, usdBalance: action.usdBalance})
    case SAVE_FUEL_BALANCE:
      return assocIdentity(state, action.address, 'fuel', action.balance)
    case SAVE_IPFS_PROFILE:
      return assocIdentity(state, action.address, 'ipfsProfile', action.ipfsHash)
    case REMOVE_IDENTITY:
      return updateIn(get(state, 'currentIdentity') === action.address ? dissoc(state, 'currentIdentity') : state, ['byAddress'], (identities) => dissoc(identities, action.address))
    case STORE_EXTERNAL_UPORT:
      return updateIn(state, ['external', action.address], (profile) => merge(profile, toClj(action.profile)))
    case STORE_ACTIVITY:
      return assocIn(state, ['byAddress', action.activity.target, 'activities', action.activity.id.toString()], toClj(action.activity))
    case UPDATE_ACTIVITY:
      activityPair = findActivity(state, action.activityId)
      if (activityPair) {
        return assocIn(state, ['byAddress', activityPair.address, 'activities', action.activityId], merge(activityPair.activity, toClj(action.changes)))
      } else return state
    case OPEN_ACTIVITY:
      activityPair = findActivity(state, action.activityId)
      if (activityPair) {
        return assocIn(state, ['byAddress', activityPair.address, 'activities', action.activityId], assoc(activityPair.activity, 'opened', true))
      } else return state
    case REMOVE_ACTIVITY:
      activityPair = findActivity(state, action.activityId)
      if (activityPair) {
        return updateIn(state, ['byAddress', activityPair.address, 'activities'], (activities) => dissoc(activities, action.activityId))
      } else return state
    case UPDATE_INTERACTION_STATS:
      return updateIn(state, ['byAddress', action.address, 'stats', action.party, action.interactionType], inc)
    case STORE_ATTESTATION:
      const tokenHash = sha3_256(action.attestation.token)
      return assocIn(state, ['byAddress', action.attestation.sub, 'attestations', tokenHash], toClj(action.attestation))
    case REMOVE_ATTESTATION:
      return updateIn(state, ['byAddress', action.address, 'attestations'], (attestations) => dissoc(attestations, action.attestationHash))
    case STORE_CURRENT_IDENTITY:
      return assoc(state, 'currentStoredIdentity', action.address)
    case STORE_KEYCHAIN_LEVEL:
      return assoc(state, 'keychainSecurityLevel', action.level)
    case STORE_SECURITY_LEVEL:
      return assocIdentity(state, action.address, 'securityLevel', action.level)
    case IMPORT_SNAPSHOT:
      const snapshot = toClj(action.snapshot.uport)
      if (snapshot && get(snapshot, 'byAddress')) {
        return massage(snapshot)
      } else {
        return state
      }
    default:
      return state
  }
}

export default uportReducer
