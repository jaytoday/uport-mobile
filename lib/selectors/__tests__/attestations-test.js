// Copyright (C) 2018 ConsenSys AG
//
// This file is part of uPort Mobile App.
//
// uPort Mobile App is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// uPort Mobile App is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with uPort Mobile App.  If not, see <http://www.gnu.org/licenses/>.
//
import uportReducer from 'uPortMobile/lib/reducers/uportReducer'
import { storeIdentity } from 'uPortMobile/lib/actions/uportActions'
import {
  attestationLedger,
  onlyLatestAttestations,
  requestedClaims,
  missingClaims,
  attestationSummary,
  allClaims,
  verifiedClaimsTokens,
  thirdPartyAttestationSummary,
  onlyLatestAttestationsWithIssuer,
  hasAttestations,
  requestedOwnClaims
} from 'uPortMobile/lib/selectors/attestations'
import { toJs, toClj } from 'mori'

const simpleIdentity = {
  address: 'did:ethr:0x9df0e9759b17f34e9123adbe6d3f25d54b72ad6a',
  own: {
    name: 'Alice',
    phone: '555-555-5555',
    description: 'UX Designer',
    connections: {
      knows: ['did:ethr:0x1bc5cbf71b068642fc5e89c5e3d0bdb3e366bc01']
    }
  },
  attestations: {
    '0011': {
      iss: 'did:ethr:0x1bc5cbf71b068642fc5e89c5e3d0bdb3e366bc01',
      iat: 1555183565000,
      exp: 1555183566000,
      claim: {
        name: 'Alice'
      },
      token: '0011.TOKEN'
    },
    '0015': {
      iss: 'did:ethr:0x1bc5cbf71b068642fc5e89c5e3d0bdb3e366bc01',
      iat: 1555183566,
      exp: 1555183567,
      claim: {
        name: 'Alice'
      },
      token: '0015.TOKEN'
    },
    '0012': {
      iat: 1555183567,
      iss: 'did:ethr:0x1bc5cbf71b068642fc5e89c5e3d0bdb3e366bc02',
      claim: {
        name: 'Alice'
      },
      token: '0012.TOKEN'
    },
    '0013': {
      iat: 1555183568,
      iss: 'did:ethr:0x1bc5cbf71b068642fc5e89c5e3d0bdb3e366bc01',
      claim: {
        phone: '555-555-5555'
      },
      token: '0013.TOKEN'
    },
    '0014': {
      iat: 1555183569,
      iss: 'did:ethr:0x1bc5cbf71b068642fc5e89c5e3d0bdb3e366bc01',
      claim: {
        nationalId: 'xxx0xxxx0xxx'
      },
      token: '0014.TOKEN'
    }
  }
}

const withoutAttestations = toClj({
  address: 'did:ethr:0x9df0e9759b17f34e9123adbe6d3f25d54b72ad6a'
})

const withIssuers = toClj({
  external: {
    'did:ethr:0x1bc5cbf71b068642fc5e89c5e3d0bdb3e366bc01': {
      name: 'Issuer 1'
    },
    'did:ethr:0x1bc5cbf71b068642fc5e89c5e3d0bdb3e366bc02': {
      name: 'Issuer 2'
    }
  }
})
const uport = uportReducer(withIssuers, storeIdentity(simpleIdentity))
const withIdentity = {uport}

const otherIdentity = {
  network: 'kovan',
  address: '34ukSmiK1oA1C5Du8aWpkjFGALoH7nsHeDX',
  own: {
    name: 'Alice Kovan',
    phone: '555-555-5555'
  },
  attestations: {
    '0020': {
      iss: 'did:ethr:0x1bc5cbf71b068642fc5e89c5e3d0bdb3e366bc01',
      iat: 1555183568,
      claim: {
        phone: '555-555-5555'
      },
      token: '0013.TOKEN'
    },
    '0021': {
      iss: 'did:ethr:0x1bc5cbf71b068642fc5e89c5e3d0bdb3e366bc01',
      iat: 1555183569,
      claim: {
        nationalId: 'xxx0xxxx0xxx'
      },
      token: '0014.TOKEN'
    }
  }
}

const withMultipleIdentities = {
  uport: uportReducer(uport, storeIdentity(otherIdentity))
}

it('returns full ledger', () => {
  expect(toJs(attestationLedger({}))).toMatchSnapshot()
  expect(toJs(attestationLedger(withIdentity))).toMatchSnapshot()
  expect(toJs(attestationLedger(withMultipleIdentities))).toMatchSnapshot()
})

it('returns ledger with only the latest attestations', () => {
  expect(toJs(onlyLatestAttestations({}))).toEqual([])
  expect(toJs(onlyLatestAttestations(withIdentity))).toMatchSnapshot()
  expect(toJs(onlyLatestAttestations(withMultipleIdentities))).toMatchSnapshot()
})

it('returns third party attestations summary', () => {
  expect(toJs(thirdPartyAttestationSummary(withIdentity))).toMatchSnapshot()
  expect(toJs(thirdPartyAttestationSummary(withMultipleIdentities))).toMatchSnapshot()
})

it('returns full attestations summary', () => {
  expect(attestationSummary(withIdentity)).toMatchSnapshot()
  expect(attestationSummary(withMultipleIdentities)).toMatchSnapshot()
})

it('returns allClaims', () => {
  expect(toJs(allClaims(withIdentity))).toMatchSnapshot()
  expect(toJs(allClaims(withMultipleIdentities))).toMatchSnapshot()
})

describe('requestedClaims', () => {
  it('returns self attested', () => {
    expect(toJs(requestedClaims(withIdentity, {}))).toEqual({})
    expect(toJs(requestedClaims(withIdentity, {requested: ['name', 'phone']}))).toEqual({name: 'Alice', phone: '555-555-5555'})
    expect(toJs(requestedClaims(withIdentity, {requested: ['name', 'phone'], verified: ['nationalId', 'nonexisting']}))).toEqual({name: 'Alice', nationalId: 'xxx0xxxx0xxx', phone: '555-555-5555'})
  
    expect(toJs(requestedClaims(withMultipleIdentities, {}))).toEqual({})
    expect(toJs(requestedClaims(withMultipleIdentities, {requested: ['name']}))).toEqual({name: 'Alice Kovan'})
    expect(toJs(requestedClaims(withMultipleIdentities, {requested: ['name', 'phone'], verified: ['nationalId', 'nonexisting']}))).toEqual({name: 'Alice Kovan', nationalId: 'xxx0xxxx0xxx', phone: '555-555-5555'})
  })

  it('handles claims specs', () => {
    expect(toJs(requestedClaims(withIdentity, {claims: {user_info: {name: null}}}))).toEqual({name: 'Alice'})
    expect(toJs(requestedClaims(withIdentity, {claims: {verifiable: {name: {iss: ['did:ethr:0x1bc5cbf71b068642fc5e89c5e3d0bdb3e366bc01']}}}}))).toEqual({name: 'Alice'})
    expect(toJs(requestedClaims(withIdentity, {claims: {verifiable: {name: {iss: 'did:ethr:0x1bc5cbf71b068642fc5e89c5e3d0bdb3e366bc01'}, nationalId: null}}}))).toEqual({name: 'Alice', nationalId: 'xxx0xxxx0xxx'})
    expect(toJs(requestedClaims(withIdentity, {claims: {user_info: {name: null}, verifiable: {nationalId: null}}}))).toEqual({name: 'Alice', nationalId: 'xxx0xxxx0xxx'})
  })

  it('ignores old specs if claims spec exists', () => {
    expect(toJs(requestedClaims(withIdentity, {requested: ['description'], verified: ['nationalId'], claims: {user_info: {name: null}}}))).toEqual({name: 'Alice', nationalId: 'xxx0xxxx0xxx'})
    expect(toJs(requestedClaims(withIdentity, {requested: ['description'], verified: ['nationalId'], claims: {verifiable: {name: {iss: ['did:ethr:0x1bc5cbf71b068642fc5e89c5e3d0bdb3e366bc01']}}}}))).toEqual({name: 'Alice', description: 'UX Designer'})
    expect(toJs(requestedClaims(withIdentity, {requested: ['description'], verified: ['nationalId'], claims: {user_info: {phone: null}, verifiable: {name: null}}}))).toEqual({name: 'Alice', phone: '555-555-5555'})
  })
})

describe('requestedOwnClaims', () => {
  describe('openid like', () => {
    it('returns self attested', () => {
      expect(toJs(requestedOwnClaims(withIdentity, undefined))).toEqual({})
      expect(toJs(requestedOwnClaims(withIdentity, {}))).toEqual({})
      expect(toJs(requestedOwnClaims(withIdentity, {claims: {user_info: {}}}))).toEqual({})
      expect(toJs(requestedOwnClaims(withIdentity, {claims: {user_info: {name: null}}}))).toEqual({name: 'Alice'})
      expect(toJs(requestedOwnClaims(withIdentity, {requested: ['name'], claims: {}}))).toEqual({name: 'Alice'})
      expect(toJs(requestedOwnClaims(withIdentity, {claims: {user_info: {name: {essential: true}, nationalId: null, nonexisting: null}}}))).toEqual({name: 'Alice'})
      expect(toJs(requestedOwnClaims(withMultipleIdentities, {claims: {user_info: {}}}))).toEqual({})
      expect(toJs(requestedOwnClaims(withMultipleIdentities, {claims: {user_info: {name: null}}}))).toEqual({name: 'Alice Kovan'})
      expect(toJs(requestedOwnClaims(withMultipleIdentities, {claims: {user_info: {name: {essential: true}, nationalId: null, nonexisting: null}}}))).toEqual({name: 'Alice Kovan'})
    })
  })

  describe('LEGACY', () => {
    it('returns self attested', () => {
      expect(toJs(requestedOwnClaims(withIdentity, {requested: []}))).toEqual({})
      expect(toJs(requestedOwnClaims(withIdentity, {requested: ['name']}))).toEqual({name: 'Alice'})
      expect(toJs(requestedOwnClaims(withIdentity, {requested: ['name', 'nationalId', 'nonexisting']}))).toEqual({name: 'Alice'})
      expect(toJs(requestedOwnClaims(withMultipleIdentities, {requested: []}))).toEqual({})
      expect(toJs(requestedOwnClaims(withMultipleIdentities, {requested: ['name']}))).toEqual({name: 'Alice Kovan'})
      expect(toJs(requestedOwnClaims(withMultipleIdentities, {requested: ['name', 'nationalId', 'nonexisting']}))).toEqual({name: 'Alice Kovan'})
    })
  })
})

describe('missingClaims', () => {
  it('handles missing claims specs', () => {
    expect(missingClaims(withIdentity, {claims: {verifiable: {name: {iss: [{did: 'did:ethr:0x1bc5cbf71b068642fc5e89c5e3d0bdb3e366bc01'}]}}}})).toEqual([])
    expect(missingClaims(withIdentity, {claims: {verifiable: {cityId: {iss: [{did: 'did:web:city.claims', url: 'https://city.claims'}], essential: true, reason: 'We need this'}, name: null}}})).toEqual([{claimType: 'cityId', iss: [{did: 'did:web:city.claims', url: 'https://city.claims'}], essential: true, reason: 'We need this'}])
    expect(missingClaims(withIdentity, {claims: {verifiable: {name: {iss: [{did: 'did:ethr:0x1bc5cbf71b068642fc5e89c5e3d0bdb3e366bc0a'}]}}}})).toEqual([{claimType: 'name', iss: [{did: 'did:ethr:0x1bc5cbf71b068642fc5e89c5e3d0bdb3e366bc0a'}]}])
    expect(missingClaims(withIdentity, {claims: {verifiable: {name: {iss: [{did: 'did:ethr:0x1bc5cbf71b068642fc5e89c5e3d0bdb3e366bc0b'}]}, nationalId: null}}})).toEqual([{claimType: 'name', iss: [{did: 'did:ethr:0x1bc5cbf71b068642fc5e89c5e3d0bdb3e366bc0b'}]}])
  })

  it('LEGACY FORMAT', () => {
    expect(missingClaims(withIdentity, {verified: ['name']})).toEqual([])
    expect(missingClaims(withIdentity, {verified: ['name', 'cityId']})).toEqual([{claimType: 'cityId'}])
  })
})

describe('verifiedClaimsTokens', () => {
  it('returns verifiedClaimsTokens', () => {
    expect(toJs(verifiedClaimsTokens(withIdentity, {}))).toEqual([])
    expect(toJs(verifiedClaimsTokens(withIdentity, {claims: {verifiable: { name: null }}}))).toEqual(['0012.TOKEN', '0015.TOKEN'])
    expect(toJs(verifiedClaimsTokens(withIdentity, {claims: {verifiable: { name: null, nationalId: null, nonexisting: null }}}))).toEqual(['0012.TOKEN', '0015.TOKEN', '0014.TOKEN'])

    expect(toJs(verifiedClaimsTokens(withMultipleIdentities, {}))).toEqual([])
    expect(toJs(verifiedClaimsTokens(withMultipleIdentities, {claims: {verifiable: { name: null }}}))).toEqual([])
    expect(toJs(verifiedClaimsTokens(withMultipleIdentities, {claims: {verifiable: { name: null, nationalId: null, nonexisting: null }}}))).toEqual(['0014.TOKEN'])
  })

  it('handles required issuers', () => {
    expect(toJs(verifiedClaimsTokens(withIdentity, {claims: {verifiable: { name: { iss: [{did: 'did:ethr:0x1bc5cbf71b068642fc5e89c5e3d0bdb3e366bc01'}] } }}}))).toEqual(['0015.TOKEN'])
    expect(toJs(verifiedClaimsTokens(withIdentity, {claims: {verifiable: { name: { iss: [{did: 'did:ethr:0x1bc5cbf71b068642fc5e89c5e3d0bdb3e366bc01'}]}, nationalId: null }}}))).toEqual(['0015.TOKEN', '0014.TOKEN'])
  })

  it('LEGACY FORMAT', () => {
    expect(toJs(verifiedClaimsTokens(withIdentity, {verified: ['name']}))).toEqual(['0012.TOKEN', '0015.TOKEN'])
    expect(toJs(verifiedClaimsTokens(withIdentity, {verified: ['name', 'nationalId', 'nonexisting']}))).toEqual(['0012.TOKEN', '0015.TOKEN', '0014.TOKEN'])

    expect(toJs(verifiedClaimsTokens(withMultipleIdentities, {verified: ['name']}))).toEqual([])
    expect(toJs(verifiedClaimsTokens(withMultipleIdentities, {verified: ['name', 'nationalId', 'nonexisting']}))).toEqual(['0014.TOKEN'])
  })

  it('ignores legacy format if openid like format extists', () => {
    expect(toJs(verifiedClaimsTokens(withIdentity, {verified: ['nationalId'], claims: {verifiable: { name: null }}}))).toEqual(['0012.TOKEN', '0015.TOKEN'])
  })
})

it('returns all onlyLatestAttestationsWithIssuer', () => {
  expect(onlyLatestAttestationsWithIssuer(withIdentity)).toMatchSnapshot()
  expect(onlyLatestAttestationsWithIssuer(withMultipleIdentities)).toMatchSnapshot()
})

describe('hasAttestations', () => {
  expect(hasAttestations(withIdentity)).toBeTruthy()
  expect(hasAttestations(withoutAttestations)).toBeFalsy()
})
