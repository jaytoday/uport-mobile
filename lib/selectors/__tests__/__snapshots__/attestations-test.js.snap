// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`returns all onlyLatestAttestationsWithIssuer 1`] = `
Array [
  Object {
    "claim": Object {
      "nationalId": "xxx0xxxx0xxx",
    },
    "iat": 1555183569,
    "iss": "did:ethr:0x1bc5cbf71b068642fc5e89c5e3d0bdb3e366bc01",
    "issuer": Object {
      "name": "Issuer 1",
    },
    "token": "0014.TOKEN",
  },
  Object {
    "claim": Object {
      "phone": "555-555-5555",
    },
    "iat": 1555183568,
    "iss": "did:ethr:0x1bc5cbf71b068642fc5e89c5e3d0bdb3e366bc01",
    "issuer": Object {
      "name": "Issuer 1",
    },
    "token": "0013.TOKEN",
  },
  Object {
    "claim": Object {
      "name": "Alice",
    },
    "iat": 1555183567,
    "iss": "did:ethr:0x1bc5cbf71b068642fc5e89c5e3d0bdb3e366bc02",
    "issuer": Object {
      "name": "Issuer 2",
    },
    "token": "0012.TOKEN",
  },
  Object {
    "claim": Object {
      "name": "Alice",
    },
    "exp": 1555183567,
    "iat": 1555183566,
    "iss": "did:ethr:0x1bc5cbf71b068642fc5e89c5e3d0bdb3e366bc01",
    "issuer": Object {
      "name": "Issuer 1",
    },
    "token": "0015.TOKEN",
  },
]
`;

exports[`returns all onlyLatestAttestationsWithIssuer 2`] = `
Array [
  Object {
    "claim": Object {
      "nationalId": "xxx0xxxx0xxx",
    },
    "iat": 1555183569,
    "iss": "did:ethr:0x1bc5cbf71b068642fc5e89c5e3d0bdb3e366bc01",
    "issuer": Object {
      "name": "Issuer 1",
    },
    "token": "0014.TOKEN",
  },
  Object {
    "claim": Object {
      "phone": "555-555-5555",
    },
    "iat": 1555183568,
    "iss": "did:ethr:0x1bc5cbf71b068642fc5e89c5e3d0bdb3e366bc01",
    "issuer": Object {
      "name": "Issuer 1",
    },
    "token": "0013.TOKEN",
  },
]
`;

exports[`returns allClaims 1`] = `
Object {
  "description": "UX Designer",
  "name": "Alice",
  "nationalId": "xxx0xxxx0xxx",
  "phone": "555-555-5555",
}
`;

exports[`returns allClaims 2`] = `
Object {
  "name": "Alice Kovan",
  "nationalId": "xxx0xxxx0xxx",
  "phone": "555-555-5555",
}
`;

exports[`returns full attestations summary 1`] = `
Object {
  "description": Object {
    "UX Designer": 0,
  },
  "name": Object {
    "Alice": 2,
  },
  "nationalId": Object {
    "xxx0xxxx0xxx": 1,
  },
  "phone": Object {
    "555-555-5555": 1,
  },
}
`;

exports[`returns full attestations summary 2`] = `
Object {
  "name": Object {
    "Alice Kovan": 0,
  },
  "nationalId": Object {
    "xxx0xxxx0xxx": 1,
  },
  "phone": Object {
    "555-555-5555": 1,
  },
}
`;

exports[`returns full ledger 1`] = `Array []`;

exports[`returns full ledger 2`] = `
Array [
  Object {
    "claim": Object {
      "nationalId": "xxx0xxxx0xxx",
    },
    "iat": 1555183569,
    "iss": "did:ethr:0x1bc5cbf71b068642fc5e89c5e3d0bdb3e366bc01",
    "token": "0014.TOKEN",
  },
  Object {
    "claim": Object {
      "phone": "555-555-5555",
    },
    "iat": 1555183568,
    "iss": "did:ethr:0x1bc5cbf71b068642fc5e89c5e3d0bdb3e366bc01",
    "token": "0013.TOKEN",
  },
  Object {
    "claim": Object {
      "name": "Alice",
    },
    "iat": 1555183567,
    "iss": "did:ethr:0x1bc5cbf71b068642fc5e89c5e3d0bdb3e366bc02",
    "token": "0012.TOKEN",
  },
  Object {
    "claim": Object {
      "name": "Alice",
    },
    "exp": 1555183567,
    "iat": 1555183566,
    "iss": "did:ethr:0x1bc5cbf71b068642fc5e89c5e3d0bdb3e366bc01",
    "token": "0015.TOKEN",
  },
  Object {
    "claim": Object {
      "name": "Alice",
    },
    "exp": 1555183566000,
    "iat": 1555183565000,
    "iss": "did:ethr:0x1bc5cbf71b068642fc5e89c5e3d0bdb3e366bc01",
    "token": "0011.TOKEN",
  },
]
`;

exports[`returns full ledger 3`] = `
Array [
  Object {
    "claim": Object {
      "nationalId": "xxx0xxxx0xxx",
    },
    "iat": 1555183569,
    "iss": "did:ethr:0x1bc5cbf71b068642fc5e89c5e3d0bdb3e366bc01",
    "token": "0014.TOKEN",
  },
  Object {
    "claim": Object {
      "phone": "555-555-5555",
    },
    "iat": 1555183568,
    "iss": "did:ethr:0x1bc5cbf71b068642fc5e89c5e3d0bdb3e366bc01",
    "token": "0013.TOKEN",
  },
]
`;

exports[`returns ledger with only the latest attestations 1`] = `
Array [
  Object {
    "claim": Object {
      "nationalId": "xxx0xxxx0xxx",
    },
    "iat": 1555183569,
    "iss": "did:ethr:0x1bc5cbf71b068642fc5e89c5e3d0bdb3e366bc01",
    "token": "0014.TOKEN",
  },
  Object {
    "claim": Object {
      "phone": "555-555-5555",
    },
    "iat": 1555183568,
    "iss": "did:ethr:0x1bc5cbf71b068642fc5e89c5e3d0bdb3e366bc01",
    "token": "0013.TOKEN",
  },
  Object {
    "claim": Object {
      "name": "Alice",
    },
    "iat": 1555183567,
    "iss": "did:ethr:0x1bc5cbf71b068642fc5e89c5e3d0bdb3e366bc02",
    "token": "0012.TOKEN",
  },
  Object {
    "claim": Object {
      "name": "Alice",
    },
    "exp": 1555183567,
    "iat": 1555183566,
    "iss": "did:ethr:0x1bc5cbf71b068642fc5e89c5e3d0bdb3e366bc01",
    "token": "0015.TOKEN",
  },
]
`;

exports[`returns ledger with only the latest attestations 2`] = `
Array [
  Object {
    "claim": Object {
      "nationalId": "xxx0xxxx0xxx",
    },
    "iat": 1555183569,
    "iss": "did:ethr:0x1bc5cbf71b068642fc5e89c5e3d0bdb3e366bc01",
    "token": "0014.TOKEN",
  },
  Object {
    "claim": Object {
      "phone": "555-555-5555",
    },
    "iat": 1555183568,
    "iss": "did:ethr:0x1bc5cbf71b068642fc5e89c5e3d0bdb3e366bc01",
    "token": "0013.TOKEN",
  },
]
`;

exports[`returns third party attestations summary 1`] = `
Object {
  "name": Object {
    "Alice": 2,
  },
  "nationalId": Object {
    "xxx0xxxx0xxx": 1,
  },
  "phone": Object {
    "555-555-5555": 1,
  },
}
`;

exports[`returns third party attestations summary 2`] = `
Object {
  "nationalId": Object {
    "xxx0xxxx0xxx": 1,
  },
  "phone": Object {
    "555-555-5555": 1,
  },
}
`;
